components/merkledb/src/migration.rs

```diff
@@ -264,30 +399,75 @@ impl MigrationHelper {
-    /// Use [`Fork::flush_migration`] to flush the migrated data.
+    /// Use [`flush_migration`] to flush the migrated data.
     ///
-    /// [`Fork::flush_migration`]: ../struct.Fork.html#method.flush_migration
+    /// [`flush_migration`]: fn.flush_migration.html
     pub fn merge(&mut self) -> crate::Result<()> {
         let fork = mem::replace(&mut self.fork, self.db.fork());
         self.db.merge(fork.into_patch())?;
         self.fork = self.db.fork();
         Ok(())
     }
 
+    /// Executes the provided closure in a loop until all persistent iterators instantiated
+    /// within the closure have ended. After each iteration, the changes in migrated data are
+    /// merged to the database; an error is returned if this merge fails.
+    ///
+    /// If no iterators are instantiated within the closure, a single iteration will be performed.
+    pub fn iter_loop(
+        &mut self,
+        mut step: impl FnMut(&Self, &mut PersistentIters<Scratchpad<'_, &Fork>>),
+    ) -> crate::Result<()> {
+        loop {
+            let mut iterators = PersistentIters::new(self.scratchpad());
+            step(self, &mut iterators);
+            let should_break = iterators.all_ended();
+            self.merge()?;
+            if should_break {
+                break;
+            }
+        }
+        Ok(())
+    }
+
     /// Merges the changes to the migrated data to the database.
     /// Returns hash representing migrated data state, or an error if the merge has failed.
     ///
     /// `finish` does not flush the migration; the migrated data remains in a separate namespace.
-    /// Use [`Fork::flush_migration`] to flush the migrated data.
+    /// Use [`flush_migration`] to flush the migrated data.
     ///
-    /// [`Fork::flush_migration`]: ../struct.Fork.html#method.flush_migration
+    /// [`flush_migration`]: fn.flush_migration.html
     pub fn finish(self) -> crate::Result<Hash> {
         let patch = self.fork.into_patch();
         let hash = Migration::new(&self.namespace, &patch).state_hash();
         self.db.merge(patch).map(|()| hash)
     }
 }
 
+/// Flushes the migration to the fork. Once the `fork` is merged, the migration is complete.
+///
+/// The following operations will be performed:
+///
+/// - Migrated indexes will replace their old versions
+/// - Migrated indexes will be aggregated in the default namespace
+/// - Indexes marked with tombstones will be removed
+/// - Scratchpad associated with the migration will be cleared
+pub fn flush_migration(fork: &mut Fork, namespace: &str) {
+    fork.flush_migration(namespace);
+    Scratchpad::new(namespace, &*fork).clear();
+}
+
+/// Rolls back the migration.
+///
+/// The following operations will be performed:
+///
+/// - Migrated indexes will be erased (both data and metadata)
+/// - Scratchpad associated with the migration will be cleared
+pub fn rollback_migration(fork: &mut Fork, namespace: &str) {
```

Comment:
There is no concept of "unknown" migrations / scratchpads right now. Similar to indexes, migrations / scratchpads are created on first use. Should this be added to the docs?

Reply:
Yeah, I think so. And one more question: are we going to change this in the future? Unlike indexes, I can't imagine a scenario where implicit migration creation is useful.

Reply:
I think we can, but since the main crate is supposed to hide migration initialization details from the service developers, we can defer this decision for a little.